lib/logstash/inputs/redis.rb

```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
     end
   end
 
+  # private
+  def list_pattern_stop
+    return if @redis.nil? || !@redis.connected?
+
+    @redis.quit rescue nil
+    @redis = nil
+    @threadpool.shutdown
+    @threadpool.wait_for_termination
+  end
+
+  # private
+  def worker_consume(output_queue, key)
+    @logger.debug("Launched worker for #{key}")
+    redis = new_redis_instance
+    begin
+      (0..@pattern_list_max_items-1).each do
+        if stop?
+          @logger.debug("Breaking from thread #{key} as it was requested to stop")
+          break
+        end
+        value = redis.lpop(key)
+        break if value.nil?
+        queue_event(value, output_queue)
+      end
+    ensure
+      redis.quit rescue nil
+    end
+    @logger.debug("Exit worker for #{key}")
+  end
+
+  # private
+  def threadpool_capacity?
+    @threadpool.remaining_capacity > 0
+  end
+
+  # private
+  def launch_worker(output_queue, key)
+    @current_workers.add(key)
+    @threadpool.post do
+      begin
+        worker_consume(output_queue, key)
+      ensure
+        @current_workers.delete(key)
+      end
+    end
+  end
+
+  # private
+  def ensure_workers(output_queue)
+    return unless threadpool_capacity?
+    keys = @redis.keys(@key)
+    keys.shuffle
+    keys.each do |key|
+      next if @current_workers.include?(key)
+      launch_worker(output_queue, key)
+      break unless threadpool_capacity?
+    end
+  end
+
+  # private
+  def init_list_pattern_runner
+    init_threadpool
+    @redis ||= connect
```

Comment:
Should we use `redis_runner` here as it seems to handle failures and reconnection? As a general thing seeing the `@redis` usage in the PR changes, I'd try to avoid touching the `@redis` instance variable directly as much as possible since it seems to be used by the other internal logic and anything we do here for our own purposes can easily conflict (or we end up doing the same things again).

Reply:
Good catch. The starting point for this new logic was the `list_runner` logic which is pretty different from the pubsub stuff. But could definitely use redis_runner for the main loop.